BUG: Fix ticklabels getter in `CoordinateHelper`

Carifio24 opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
https://docs.astropy.org/en/latest/development/quickstart.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

While working with some code that references  the `ticklabels` attribute of `CoordinateHelper` (which I realize needs to be updated as these properties are deprecated), I noticed that the `ticklabels` getter returns the ticks rather than the labels. This is a small PR to fix that. This previously worked but looks to have been broken in astropy 7 (in particular in https://github.com/astropy/astropy/commit/1425e1c72ee679c8088759fa639355fd4dab8c70).

I'm happy to add a test for these setters (something that checks the type of the returned value?) if that's desired.

<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->

<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

neutrinoceros commented:
Hi @Carifio24, thanks for working on this.
Without being too familiar with the broad context of this class, I'm surprised to see some deprecated code in need for change; deprecated code should be expected to stay untouched until removal.
That said, it is possible that something genuinely broke following an oversight in #16685, but I would need some more context to call it. Adding a test would both help show case what's broken and prevent another regression in the future, so please add one.

Carifio24 commented:
@neutrinoceros Agreed that it's an unusual situation. Basically what happened was:
* `CoordinateHelpers` properties that had public were changed to private
* Public getters and setters were added for these properties, with a deprecation warning. The bug in question is that one of these getters returns the wrong properties

So while the properties are deprecated, the code adding the getters/setters is relatively new. I've added a test to check that these getters return the appropriate values; happy to revise as needed.

astrofrog reviewed (changes requested):
The fix is good but I think the test isn't quite right - I don't think we should access private properties in tests. I would just test the three deprecated tick/ticklabel/axislabel by accessing methods on them, and calling the test 'test_deprecated_getters' and we can remove the test in 8.0.

Carifio24 commented:
@astrofrog I've updated the test as you suggested. I don't know the details of these classes super well so I just added a few simple asserts that would fail if the getter returned the wrong type, since that's the issue being tested here.

neutrinoceros commented on astropy/visualization/wcsaxes/tests/test_coordinate_helpers.py:
```diff
@@ -163,3 +165,15 @@ def test_get_position():
+
+
+def test_deprecated_getters():
+    fig = plt.figure()
```
please use `plt.subplots()` instead. `plt.figure`; the latter isn't thread safe and may create hard-to-debug situations when the test suite is run with multi-threading.

neutrinoceros commented on astropy/visualization/wcsaxes/tests/test_coordinate_helpers.py:
```diff
@@ -163,3 +165,15 @@ def test_get_position():
+    ax = WCSAxes(fig, [0.1, 0.1, 0.8, 0.8], aspect="equal")
+    helper = CoordinateHelper(parent_axes=ax)
+
+    with warnings.catch_warnings():
```
are all 3 `assert` statement expected to emit a warning ? in general, we shouldn't be filtering these in test, but instead use `pytest.warns` to mark these warnings as *expected*. It's also best to keep the code under such contexts a minimal as possible, so only the lines that do emit the warning exist in that block.

astrofrog replied:
Yes I was going to make the same comment - we should have a separate pytest.warns context for each of the three lines to make sure we are emitting the deprecation warning.

Carifio24 replied:
Yes, each of the three `assert` statements should emit. I've put each one into a separate `pytest.warns` context as suggested.

neutrinoceros commented on astropy/visualization/wcsaxes/tests/test_coordinate_helpers.py:
```diff
@@ -163,3 +164,16 @@ def test_get_position():
+    helper = CoordinateHelper(parent_axes=ax)
+
+    with pytest.warns(AstropyDeprecationWarning):
+        assert not helper.ticks.get_display_minor_ticks()
```
```suggestion
        ticks = helper.ticks
    assert not ticks.get_display_minor_ticks()
```

neutrinoceros commented on astropy/visualization/wcsaxes/tests/test_coordinate_helpers.py:
```diff
@@ -163,3 +164,16 @@ def test_get_position():
+    with pytest.warns(AstropyDeprecationWarning):
+        assert helper.ticklabels.text == {}
+    with pytest.warns(AstropyDeprecationWarning):
+        assert helper.axislabels.get_visibility_rule() == "labels"
```
```suggestion
        axislabels = helper.axislabels
    assert axislabels.get_visibility_rule() == "labels"
```

neutrinoceros reviewed (commented):
almost there ! here's a couple suggestions to reduce the scope of `pytest.warns` to the bare minimum

Carifio24 commented:
@neutrinoceros I've updated the test to reduce the scope of `pytest.warns` as suggested.

neutrinoceros commented on astropy/visualization/wcsaxes/tests/test_coordinate_helpers.py:
```diff
@@ -163,3 +164,19 @@ def test_get_position():
+    assert not ticks.get_display_minor_ticks()
+    with pytest.warns(AstropyDeprecationWarning):
+        ticklabels = helper.ticklabels
+    assert ticklabels.text == {}
```
Just to be thorough: I just checked how this tested behaved when run without the companion patch, and this assert is the only one that fails. Is it what you expect ?

Carifio24 replied:
Yeah, that's what I would expect. I was following Tom's suggestion of testing the three deprecated getters, but only the `ticklabels` getter is broken right now. The checks themselves aren't particularly meaningful beyond the fact that if the getter returns the wrong member (as happens with `ticklabels`) the statement should fail.

astrofrog reviewed (commented):
Looks good! One last request, can you add a bugfix changelog entry? (in ``docs/changes``)

neutrinoceros reviewed (approved):
approving now, modulo Tom's request for a changelog entry

Carifio24 commented:
Bugfix changelog entry added!

neutrinoceros commented on docs/changes/visualization/17444.bugfix.rst:
```diff
@@ -0,0 +1 @@
+Fix broken `ticklabels` getter in `CoordinateHelper`. The getter was incorrectly returning the helper's ticks rather than the labels; this PR corrects that.
```
The context of "this PR" will be lost once the changelog is assembled.
```suggestion
Fix ``CoordinateHelper.ticklabels``. The getter was incorrectly returning
the helper's ticks rather than the labels.
```
